### PR TITLE
Add Optuna support for HPO

### PR DESCRIPTION
**Pull request by shihchengli**
## Description
Address #783. As discussed in PR #734, raytune + optuna couldn't handle the nested distribution search space such as below.
`"dropout": tune.choice([tune.choice([0.0]), tune.quniform(lower=0.05, upper=0.4, q=0.05)])`
To bypass this issue, zeros are repeated multiple times alongside the other possible values in the `tune.choice`.

## Relevant issues
#783

## Checklist
- [ ] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

File: tests/cli/test_cli_regression_mol.py
Comment (hwpang) [inline tests/cli/test_cli_regression_mol.py:367]:
```suggestion
@pytest.mark.skipif(NO_RAY or NO_OPTUNA, reason="Optuna not installed")
```

File: tests/cli/test_cli_regression_mol.py
Comment (hwpang) [inline tests/cli/test_cli_regression_mol.py:396]:
```suggestion
@pytest.mark.skipif(NO_RAY or NO_HYPEROPT, reason="Ray and/or Hyperopt not installed")
```

File: .github/workflows/ci.yml
Comment (hwpang) [inline .github/workflows/ci.yml:87]: I think maybe we can provide two possible installation for hpopt, one comes with hyperopt and the other optuna, so that the user can use `pip install ".[hpopt-hyperopt]"` or `pip install ".[hpopt-optuna]"` when they install the hpopt module. What do you think? It doesn't have to be `hpopt-hyperopt` and `hpopt-optuna`, but something along those lines.

**Review by hwpang — CHANGES_REQUESTED** [review]
LGTM! I only have some minor requests for changes.

File: .github/workflows/ci.yml
Comment (shihchengli) [reply .github/workflows/ci.yml:87]: I think putting everything into `hpopt` should be fine (i.e., `hpopt = ["ray[tune]", "hyperopt", "optuna"]`). The only difference between `hpopt-hyperopt` and `hpopt-optuna` is whether to install `hyperopt` or `optuna`. If users prefer not to install the additional dependency of either one, they can manually install the packages they want using pip and I don't think we need to provide the additional installation choices for this.

**Comment by akshatzalte** [comment]
LGTM

**Review by akshatzalte — APPROVED** [review]
LGTM

**Comment by shihchengli** [comment]
@hwpang let me know if you are fine with installing Optuna when `pip install ".[hpopt]"` is used. I will merge this PR after your approval.

**Review by hwpang — APPROVED** [review]
LGTM

**Pull request merged**
